### PR TITLE
[FIX] point_of_sale: apply rewards on gs1 barcode scan

### PR DESCRIPTION
**Pull request by lase-odoo**
### Steps to reproduce:

- In the settings:
  - Change the Barcode Nomenclature to Default gs1
  - Enable Promotions, Coupons, ... on Point of sale
- Create a prodcut with barcode 95412427100283
- Point of Sale > Product > Discount & Loyalty
- Create a new discount and change the conditional rule to set your product and a min purchase of 0
- Open a new pos session
- Scan the gs1 barcode 0195412427100283

### Expected behavior:

Just as if you scanned the barcode 95412427100283 directly the discount should be applied.

Cause of the issue:

When a barcode is scanned in the pos, the ` _scan` method is called to parse the barcode and to trigger the associated action:
https://github.com/odoo/odoo/blob/775827b8f7fa95f2afd77a9b42d6cd8e436ea690/addons/point_of_sale/static/src/app/barcode/barcode_reader_service.js#L69-L79
This flow use to not update the rewards after a product scanned and was fixed by commit 21392ab when the barcode associated action is `_barcodeProductAction` because fo these lines:
https://github.com/odoo/odoo/blob/775827b8f7fa95f2afd77a9b42d6cd8e436ea690/addons/pos_loyalty/static/src/overrides/components/product_screen/product_screen.js#L105-L107 however, if the code is identified to be gs1
Howecver, if the barcode is identified to be of the gs1 nomenclature it will trigger the `_barcodeGS1Action` which still does not update the rewards.

### Fix:

Inspired by commit 21392ab

opw-3813858
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by robinengels — APPROVED** [review]
Amazing fix thank you !

File: addons/point_of_sale/static/src/app/barcode/barcode_reader_service.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 /** @odoo-module **/
+/* global posmodel */
```
Comment (vlst-odoo) [inline addons/point_of_sale/static/src/app/barcode/barcode_reader_service.js:2]: you cannot rely on `pos` here. The `barcode_reader` service is already a dependency of the `pos` service.

**Review by vlst-odoo — CHANGES_REQUESTED** [review]

File: addons/point_of_sale/static/src/app/barcode/barcode_reader_service.js (hunk 1 above)
Comment (lase-odoo) [reply addons/point_of_sale/static/src/app/barcode/barcode_reader_service.js]: @vlst-odoo, Is it better now or should I override `_barcodeGS1Action` just like in Robin's fix ?

Comment (vlst-odoo) [reply addons/point_of_sale/static/src/app/barcode/barcode_reader_service.js]: what you have now is exactly the same thing as before. You are trying to access `pos`, just in a different way. The barcode service is meant to be generic, so it shouldn't reference pos. This would still be the case even if we didn't have the circular dependency problem. 

Pos loyalty doesn't have anything to do about the barcode scanner. The barcode scanner just tells you what barcodes were scanned, it shouldn't perform side effects. 

Asking the barcode scanner to worry about pos loyalty is a bit like asking your mailman to also pay for the invoices that he delivers to you. In principle you could work out an arrangement with him, but that doesn't really work :)

In short, you are right, you should either override one of the existing listeners ( like `_barcodeGS1Action` ) or maybe it would be even better to directly call `useBarcodeReader` again in `pos_loyalty`

Comment (lase-odoo) [reply addons/point_of_sale/static/src/app/barcode/barcode_reader_service.js]: Thanks for the review and the clear explanation @vlst-odoo I updated the diff :slightly_smiling_face:

**Comment by lase-odoo** [comment]
@vlst-odoo should I change something else ? :slightly_smiling_face:

**Comment by vlst-odoo** [comment]
> @vlst-odoo should I change something else ? 🙂

We can merge when the runbot is green

**Comment by lase-odoo** [comment]
@vlst-odoo It is ready then 🫡

**Review by vlst-odoo — APPROVED** [review]
robodoo r+
